pkg/operation/botanist/operatingsystemconfig.go

```diff
@@ -269,13 +270,23 @@ func (b *Botanist) generateCloudConfigExecutorResourcesForWorker(
 
 	executorScript, err := ExecutorScriptFn([]byte(oscDataOriginal.Content), hyperkubeImage, kubernetesVersion.String(), kubeletDataVolume, *oscDataOriginal.Command, oscDataOriginal.Units)
 	if err != nil {
-		return "", nil, err
+		return nil, nil, err
 	}
 
-	resources, err := registry.AddAllAndSerialize(executor.Secret(secretName, metav1.NamespaceSystem, worker.Name, executorScript))
-	if err != nil {
-		return "", nil, err
+	if err := registry.Add(executor.Secret(secretName, metav1.NamespaceSystem, worker.Name, executorScript)); err != nil {
+		return nil, nil, err
+	}
+
+	// TODO(rfranzke): Remove this legacySecretName in a future release.
+	// Since we changed the logic of the Key function to incorporate the CRI name for
+	// https://github.com/gardener/gardener/issues/4415, the name of the secret changes. To make existing nodes
+	// update to the new secret name, we need to also create the legacy secret name and point it to the new secret.
+	if legacySecretName := operatingsystemconfig.Key(worker.Name, kubernetesVersion, nil); legacySecretName != secretName {
```

Comment:
I don't understand this if check.
Isn't this always `true` if any cri is configured? So also for newly created nodes we create a second cloud-config downloader secret with the old name?
Or what am I missing here?

Reply:
The `legacySecretName` is different from `secretName` only when the worker pool does use `containerd`. If the worker pool has `.cri=nil` or `.cri.name=docker` then `legacySecretName=secretName`. This is check is to prevent adding the same secret twice to the `ManagedResource`.